base_layer/core/src/validation/chain_validators.rs

```diff
@@ -0,0 +1,150 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+pub use crate::consensus::ConsensusManager;
+use crate::{
+    blocks::{
+        blockheader::{BlockHeader, BlockHeaderValidationError},
+        genesis_block::get_gen_block_hash,
+    },
+    chain_storage::{BlockchainBackend, BlockchainDatabase, MmrTree},
+    validation::{error::ValidationError, traits::Validation},
+};
+use std::sync::Arc;
+use tari_crypto::commitment::HomomorphicCommitmentFactory;
+use tari_transactions::types::CryptoFactories;
+use tari_utilities::hash::Hashable;
+
+/// This validator checks that the synced state satisfies *all* consensus rules and should only be performed on
+/// the chain tip header.
+pub struct ChainTipValidator<B: BlockchainBackend> {
+    rules: ConsensusManager<B>,
+    factories: Arc<CryptoFactories>,
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> ChainTipValidator<B>
+where B: BlockchainBackend
+{
+    pub fn new(rules: ConsensusManager<B>, factories: Arc<CryptoFactories>, db: BlockchainDatabase<B>) -> Self {
+        Self { rules, factories, db }
+    }
+
+    fn db(&self) -> Result<BlockchainDatabase<B>, ValidationError> {
+        Ok(self.db.clone())
+    }
+}
+
+impl<B: BlockchainBackend> Validation<BlockHeader, B> for ChainTipValidator<B> {
+    /// The consensus checks that are done (in order of cheapest to verify to most expensive):
+    /// 1. Does the MMR roots calculated from the chain state match that roots provided in the tip header?
+    /// 1. Is the full accounting balance of the current chain state valid?
+    fn validate(&self, block_header: &BlockHeader) -> Result<(), ValidationError> {
+        check_mmr_roots(block_header, self.db()?)?;
+        check_chain_accounting_balance(block_header, self.db()?, self.rules.clone(), &self.factories)?;
+
+        Ok(())
+    }
+}
+
+/// This validator checks that the synced chain builds on the correct genesis block and should only be performed on the
+/// genesis block header.
+pub struct ChainGbValidator {}
+
+impl ChainGbValidator {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl<B: BlockchainBackend> Validation<BlockHeader, B> for ChainGbValidator {
+    /// The consensus checks that are done (in order of cheapest to verify to most expensive):
+    /// 1. Is the synced genesis block header the correct block header?
+    fn validate(&self, block_header: &BlockHeader) -> Result<(), ValidationError> {
+        check_genesis_block(block_header)?;
+
+        Ok(())
+    }
+}
+
+/// Check the full accounting balance of the current chain state.
+fn check_chain_accounting_balance<B: BlockchainBackend>(
```

Comment:
This will only work at the chain tip, right?

i.e. `block_header` must be building on the tip. This won't work if you pass an older block in as `block_header`.

I think the docstring should mention this

Reply:
I extend the docstring to mention the limitation.

Reply:
👍